### PR TITLE
add CI 

### PR DESCRIPTION
**Pull request by flamerecca**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -31,6 +31,9 @@
 # Ignore all local history of files
 .history
 
+### phpstorm ###
+.idea/
+
```
Comment (shana0440) [inline .gitignore]: no idea why you add this to ignore file?
if you are using goland as IDE, please, change the phpstorm to goland.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,82 @@
+version: 2 # use CircleCI 2.0
+jobs: # basic units of work in a run
+  build: # runs not using Workflows must have a `build` job as entry point
+    docker: # run the steps with Docker
+      # CircleCI Go images available at: https://hub.docker.com/r/circleci/golang/
+      - image: circleci/golang:1.12 #
+      # CircleCI PostgreSQL images available at: https://hub.docker.com/r/circleci/postgres/
+      - image: circleci/postgres:9.6-alpine
```
Comment (shana0440) [inline .circleci/config.yml]: please remove the postgres from CI now, we haven't used it.

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        environment: # environment variables for primary container
+          POSTGRES_USER: circleci-demo-go
+          POSTGRES_DB: circle_test
+    # directory where steps are run. Path must conform to the Go Workspace requirements
+    working_directory: /go/src/github.com/CircleCI-Public/circleci-demo-go
```
Comment (shana0440) [inline .circleci/config.yml]: should not using `CircleCI-Public....`, this project not circle ci demo project, please change the working directory to this project.

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    environment: # environment variables for the build itself
+      TEST_RESULTS: /tmp/test-results # path to where test results will be saved
```
Comment (shana0440) [inline .circleci/config.yml]: can you explain why you add this environment variable?

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      - restore_cache: # restores saved cache if no changes are detected since last run
+          # Read about caching dependencies: https://circleci.com/docs/2.0/caching/
+          keys:
+            - v1-pkg-cache
+
+      # Normally, this step would be in a custom primary image;
+      # we've added it here for the sake of explanation.
+      - run: go get github.com/lib/pq
+      - run: go get github.com/mattes/migrate
+      - run: go get github.com/jstemmer/go-junit-report
```
Comment (shana0440) [inline .circleci/config.yml]: the above step seems no related to our project.

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      TEST_RESULTS: /tmp/test-results # path to where test results will be saved
+
+    steps: # steps that comprise the `build` job
+      - checkout # check out source code to working directory
+      - run: mkdir -p $TEST_RESULTS # create the test results directory
+
+      - restore_cache: # restores saved cache if no changes are detected since last run
+          # Read about caching dependencies: https://circleci.com/docs/2.0/caching/
+          keys:
+            - v1-pkg-cache
```
Comment (shana0440) [inline .circleci/config.yml]: please using `go.mod` checksum to generate the cache key, so we can avoid using the outdated cache.

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      - run:
+          name: Waiting for Postgres to be ready
+          command: |
+            for i in `seq 1 10`;
+            do
+              nc -z localhost 5432 && echo Success && exit 0
+              echo -n .
+              sleep 1
+            done
+            echo Failed waiting for Postgres && exit 1
```
Comment (shana0440) [inline .circleci/config.yml]: please remove everything related to postgres for now.

File: .circleci/config.yml [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+      - run:
+          name: Run unit tests
+          environment: # environment variables for the database url and path to migration files
+            CONTACTS_DB_URL: "postgres://circleci-demo-go@localhost:5432/circle_test?sslmode=disable"
+            CONTACTS_DB_MIGRATIONS: /go/src/github.com/CircleCI-Public/circleci-demo-go/db/migrations
+          # Store the results of our tests in the $TEST_RESULTS directory
+          command: |
+            trap "go-junit-report <${TEST_RESULTS}/go-test.out > ${TEST_RESULTS}/go-test-report.xml" EXIT
+            make test | tee ${TEST_RESULTS}/go-test.out
```
Comment (shana0440) [inline .circleci/config.yml]: as the above comment, we haven't any test case, and this seems not related to our project, so please remove it.

File: .circleci/config.yml [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+      - run: make # pull and build dependencies for the project
+
+      - save_cache: # Store cache in the /go/pkg directory
+          key: v1-pkg-cache
+          paths:
+            - "/go/pkg"
```
Comment (shana0440) [inline .circleci/config.yml]: becuase we using gomodule to manage our package, so I not sure the pacakge path is `/go/pkg` or `/go/mod/pkg`, please help to check, thanks.

File: .circleci/config.yml [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+      - run:
+          name: Start service
+          environment:
+            CONTACTS_DB_URL: "postgres://circleci-demo-go@localhost:5432/circle_test?sslmode=disable"
+            CONTACTS_DB_MIGRATIONS: /go/src/github.com/CircleCI-Public/circleci-demo-go/db/migrations
+          command: ./workdir/contacts
+          background: true # keep service running and proceed to next step
```
Comment (shana0440) [inline .circleci/config.yml]: please remove postgres and migration, it should be handle on #2

File: .circleci/config.yml [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+      - run:
+          name: Validate service is working
+          command: |
+            sleep 5
+            curl --retry 10 --retry-delay 1 -X POST --header "Content-Type: application/json" -d '{"email":"test@example.com","name":"Test User"}' http://localhost:8080/contacts
```
Comment (shana0440) [inline .circleci/config.yml]: no related to our project, please remove it.

File: .circleci/config.yml (hunk 9 above)
Comment (flamerecca) [reply .circleci/config.yml]: 這段是用來存放 cache 的
所以他不應該放在 module 的資料夾內
建議沿用他的範例寫法
